### PR TITLE
Introduce new multi-dimensional geometries

### PR DESCRIPTION
**Pull request by aprokop**
Depends on #714 for testing new geometries.

### CODE REVIEW

File: src/geometry/ArborX_HyperBox.hpp [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+ * distributed under a BSD 3-clause license. For the licensing terms see    *
+ * the LICENSE file in the top-level directory.                             *
+ *                                                                          *
+ * SPDX-License-Identifier: BSD-3-Clause                                    *
+ ****************************************************************************/
+
+#ifndef ARBORX_HYPERBOX_HPP
+#define ARBORX_HYPERBOX_HPP
+
+#if (__cplusplus >= 201703L) && (KOKKOS_VERSION >= 30699)
```
Comment (aprokop) [inline src/geometry/ArborX_HyperBox.hpp:15]: I thought about replacing `__cpluplus` check with `KOKKOS_CXX_STANDARD`, but not sure that's necessary.

File: src/geometry/ArborX_HyperBox.hpp [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+#include <ArborX_GeometryTraits.hpp>
+#include <ArborX_HyperPoint.hpp>
+
+#include <Kokkos_Macros.hpp>
+
+namespace ArborX
+{
+
+namespace Experimental
+{
```
Comment (dalg24) [inline src/geometry/ArborX_HyperBox.hpp:28]:
```suggestion
namespace ArborX::Experimental
{

```

File: src/geometry/ArborX_HyperBox.hpp [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+/**
+ * Axis-Aligned Bounding Box. This is just a thin wrapper around an array of
+ * size 2x spatial dimension with a default constructor to initialize
+ * properly an "empty" box.
+ */
+template <int DIM, class Float = float>
```
Comment (dalg24) [inline src/geometry/ArborX_HyperBox.hpp:34]:
```suggestion
template <int DIM, class FloatingPoint = float>
```
so there is no chance to confuse with plain `float`

File: src/geometry/ArborX_HyperBox.hpp [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+struct Box
+{
+private:
+  using Point = Point<DIM, Float>;
```
Comment (dalg24) [inline src/geometry/ArborX_HyperBox.hpp:38]:
```suggestion
```
too confusing

File: src/geometry/ArborX_HyperBox.hpp (hunk 1 above)
Comment (dalg24) [inline src/geometry/ArborX_HyperBox.hpp:15]: I'd prefer a
```C++
#if COND
#error MSG
#endif

File: src/geometry/ArborX_HyperBox.hpp (hunk 2 above)
Comment (aprokop) [reply src/geometry/ArborX_HyperBox.hpp:30]: There is code outside of the `Experimental`, like specialization of `GeometryTraits`.

File: src/geometry/ArborX_HyperBox.hpp (hunk 4 above)
Comment (aprokop) [reply src/geometry/ArborX_HyperBox.hpp]: I'm not opposed to changing it, but `Point` is used in 12 places inside `Box`, and it feels too verbose to spell it out.

Comment (dalg24) [reply src/geometry/ArborX_HyperBox.hpp]: Change it

Comment (aprokop) [reply src/geometry/ArborX_HyperBox.hpp]: Done.

File: src/geometry/ArborX_HyperBox.hpp (hunk 1 above)
Comment (aprokop) [reply src/geometry/ArborX_HyperBox.hpp]: Done.

File: src/geometry/ArborX_HyperBox.hpp [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+ * the LICENSE file in the top-level directory.                             *
+ *                                                                          *
+ * SPDX-License-Identifier: BSD-3-Clause                                    *
+ ****************************************************************************/
+
+#ifndef ARBORX_HYPERBOX_HPP
+#define ARBORX_HYPERBOX_HPP
+
+// FIXME_CXX17: remove first clause and update the comment once we require C++17
+#if (__cplusplus < 201703L) || (KOKKOS_VERSION < 30699)
```
Comment (masterleinad) [inline src/geometry/ArborX_HyperBox.hpp:16]: Why 3.6.99? `Kokkos` would require `C++17` for `KOKKOS_VERSION >= 3.7.99` (one the version update happened).

File: src/geometry/ArborX_HyperBox.hpp [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+template <int DIM, class FloatingPoint = float>
+struct Box
+{
+  KOKKOS_FUNCTION
+  constexpr Box()
+  {
+    for (int d = 0; d < DIM; ++d)
+    {
+      _min_corner[d] =
+          KokkosExt::ArithmeticTraits::finite_max<FloatingPoint>::value;
```
Comment (masterleinad) [inline src/geometry/ArborX_HyperBox.hpp:46]: For `3.7.00` and later `finite_min/max` are available in `Kokkos` directly.

File: src/geometry/ArborX_HyperBox.hpp (hunk 2 above)
Comment (dalg24) [reply src/geometry/ArborX_HyperBox.hpp:31]: So?  You can/should be doing `template <...> ArborX::GeometryTraits::dim<...> { ... }`

File: src/geometry/ArborX_HyperPoint.hpp [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+// FIXME_CXX17: remove first clause and update the comment once we require C++17
+#if (__cplusplus < 201703L) || (KOKKOS_VERSION < 30699)
+#error "Using multi-dimensional Point requires C++17 and Kokkos 3.7"
+#endif
+
+#include <ArborX_GeometryTraits.hpp>
+
+#include <Kokkos_Macros.hpp>
+
+#include <utility>
```
Comment (dalg24) [inline src/geometry/ArborX_HyperPoint.hpp:24]: What do we need that header include for?

Comment (aprokop) [reply src/geometry/ArborX_HyperPoint.hpp:24]: Leftover from some experimentation, removed.

File: src/geometry/ArborX_HyperBox.hpp (hunk 5 above)
Comment (aprokop) [reply src/geometry/ArborX_HyperBox.hpp:16]: There is no reason not to allow it working with Kokkos 3.7.

File: src/geometry/ArborX_HyperBox.hpp (hunk 2 above)
Comment (aprokop) [reply src/geometry/ArborX_HyperBox.hpp:31]: Done.

File: src/geometry/ArborX_HyperBox.hpp (hunk 6 above)
Comment (aprokop) [reply src/geometry/ArborX_HyperBox.hpp:46]: It's in `Kokkos::Experimental` though.

File: src/geometry/ArborX_HyperBox.hpp (hunk 5 above)
Comment (masterleinad) [reply src/geometry/ArborX_HyperBox.hpp:16]: Which features are you using here that require a newer `Kokkos` version than the minimum `ArborX` requires?

**Comment by aprokop** [comment]
I'm a bit stuck on the self-contained header tests. When we compile ArborX with C++14, we would want to exclude `ArborX_Hyper*` files from those tests, as they will error out (due to `#error` clause). However, it is not clear to me how to determine in CMake what is the standard that is used for the `ArborX` target. Does anyone know of a solution?

**Comment by masterleinad** [comment]
> I'm a bit stuck on the self-contained header tests. When we compile ArborX with C++14, we would want to exclude `ArborX_Hyper*` files from those tests, as they will error out (due to `#error` clause). However, it is not clear to me how to determine in CMake what is the standard that is used for the `ArborX` target. Does anyone know of a solution?

Isn't querying the ArborX` for https://cmake.org/cmake/help/latest/prop_tgt/CXX_STANDARD.html sufficient?

**Comment by aprokop** [comment]
> Isn't querying the ArborX` for [cmake.org/cmake/help/latest/prop_tgt/CXX_STANDARD.html](https://cmake.org/cmake/help/latest/prop_tgt/CXX_STANDARD.html) sufficient?

I tried `get_target_property(ArborXStandard ArborX CXX_STANDARD)` but it returned `NOT_FOUND`.

File: src/geometry/ArborX_HyperBox.hpp (hunk 5 above)
Comment (aprokop) [reply src/geometry/ArborX_HyperBox.hpp:16]: New geometry classes don't have volatile overloads for the reduction, which requires 3.7.

**Comment by aprokop** [comment]
For now, I could not figure out how to optionally exclude the files, so I just disabled the self-contained checks for them.

**Comment by aprokop** [comment]
It's getting tiresome pretty fast to protect the code in the tests that uses new geometries by
```
  #if (__cplusplus >= 201703L) && (KOKKOS_VERSION >= 30699)  
```
Should I introduce a new feature like `-D ArborX_USE_NEW_GEOMETRIES` which would then set requirements in the `CMakeLists.txt` to require Kokkos 3.7, set `cxx_std_17` for the `ArborX` target, and provide a macro to use?

**Comment by aprokop** [comment]
Rebased on `master` to include C++17. Also cherry-picked #714 for the testing purposes.

**Comment by aprokop** [comment]
In addition to the things implemented in this PR, we need some conversion rules for geometries.

| Source geometry type | Target geometry type | Conversion rule | Usage |
| --- | --- | --- | --- |
| `Point` | `Box` | ? | Seems unused
| `Box` | `Point` | `std::decay_t<decltype(std::declval<Box>().minCorner())>` | `check_valid_space_filling_curve`, cannot use `centroid`
| `bounding_volume` | `scene_bounding_box` | ? | `BVH` needs to know scene box type

The last one is problematic, because there is no automatic rule to do that.

**Comment by aprokop** [comment]
Another thing I stumbled upon: because I put new hyper geometries inside `Experimental`, the other things in `Experimental`, like `Ray`, `KDOP` or `Triangle` tend to pick them up first instead of `ArborX::Point` or `ArborX::Box`. Which requires a lot of explicit fixing. It could be avoided by putting them in a different namespace, but that would lead to its own problems.

**Comment by aprokop** [comment]
I think I need to add constructors to Experimental geometries to construct them from *any* point, box or sphere. So that we can mix-and-match.

**Comment by aprokop** [comment]
retest this please

**Comment by aprokop** [comment]
retest this please

File: test/tstDetailsAlgorithms.cpp [hunk 8]
```diff
@@ -15,15 +15,51 @@
...
+#include <ArborX_HyperBox.hpp>
+#include <ArborX_HyperPoint.hpp>
+#include <ArborX_HyperSphere.hpp>
+namespace ArborX
+{
+namespace ExperimentalHyperGeometry
+{
+// Deduction guide to make tests work with both old and new geometries
+template <class T>
+Point(T x, T y, T z) -> Point<3, T>;
```
Comment (dalg24) [inline test/tstDetailsAlgorithms.cpp:41]: Fine for now but eventually belongs to the `Point` header

**Review by dalg24 — APPROVED** [review]

**Pull request merged**